File: lib/typescript/codedifferently-instructional/src/multiple_choice_quiz_question.ts
import { QuizQuestion } from './quiz_question.js';
import { Scanner } from './scanner.js';

export enum AnswerChoice {
  UNANSWERED = 'UNANSWERED',
  A = 'A',
  B = 'B',
  C = 'C',
  D = 'D'
}

export class MultipleChoiceQuizQuestion extends QuizQuestion {
  private answersByOption: Map<AnswerChoice, string>;

  constructor(
    questionNumber: number,
    question: string,
    answerOptions: Map<AnswerChoice, string>,
    correctAnswer: AnswerChoice
  ) {
    super(questionNumber, question, correctAnswer);
    this.answersByOption = answerOptions;
  }

  public getAnswerChoices(): Set<AnswerChoice> {
    return new Set(this.answersByOption.keys());
  }

  public getAnswerForOption(option: AnswerChoice): string | undefined {
    return this.answersByOption.get(option);
  }

  public promptForAnswer(scanner: Scanner): void {
    let response: string | null = null;
    let answer: AnswerChoice | null = null;

    do {
      // If a response was given but not accepted, announce that.
      if (response !== null) {
        console.log("Value not accepted, try again.\n");
      }

      // Get a response.
      console.log(">> Your answer (or 's' to skip): ");
      response = scanner.next().toUpperCase().trim();

      // If the user wants to skip, let them.
      if (response === 'S') {
        this.setAnswer(null);
        return;
      }

      // Try to parse the answer.
      if (['A', 'B', 'C', 'D'].includes(response)) {
        answer = response as AnswerChoice;
      }
    } while (answer === null || !this.answersByOption.has(answer));

    // Set the question answer.
    this.setAnswer(answer);
  }

  public setAnswer(answer: AnswerChoice | null): void {
    super.setAnswer(answer ?? AnswerChoice.UNANSWERED);
  }
}

File: lib/java/codedifferently-instructional/instructional-lib/src/main/java/com/codedifferently/instructional/quiz/MultipleChoiceQuizQuestion.java
package com.codedifferently.instructional.quiz;

import java.util.Map;
import java.util.Optional;
import java.util.Scanner;
import java.util.Set;

public class MultipleChoiceQuizQuestion extends QuizQuestion {

  private Map<AnswerChoice, String> answersByOption;

  public MultipleChoiceQuizQuestion(
      int questionNumber,
      String question,
      Map<AnswerChoice, String> answerOptions,
      AnswerChoice correctAnswer) {

    super(questionNumber, question, correctAnswer.toString());
    this.answersByOption = answerOptions;
  }

  public Set<AnswerChoice> getAnswerChoices() {
    return answersByOption.keySet();
  }

  public String getAnswerForOption(AnswerChoice option) {
    return answersByOption.get(option);
  }

  @Override
  public void promptForAnswer(Scanner scanner) {
    String response = null;
    Optional<AnswerChoice> answer = Optional.empty();

    do {
      // If a response was given but not accepted, announce that.
      if (response != null) {
        System.out.println("Value not accepted, try again.\n");
      }

      // Get a response.
      System.out.print(">> Your answer (or 's' to skip): ");
      response = scanner.next().toUpperCase().trim();

      // If the user wants to skip, let them.
      if (response.equals("S")) {
        this.setAnswer(null);
        return;
      }

      // Try to parse the answer.
      answer = AnswerChoice.valueOrEmpty(response);
    } while (answer.isEmpty() || !answersByOption.containsKey(answer.get()));

    // Set the question answer.
    this.setAnswer(answer.get().toString());
  }

  @Override
  public void setAnswer(String answer) {
    super.setAnswer(answer == null ? answer : answer.toUpperCase());
  }
}
